Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
-            return;
 
-        try {
-            awaitedValue = await awaitedValue;
-            resumeMode = @GeneratorResumeModeNormal;
-        } catch (e) {
-            awaitedValue = e;
-            resumeMode = @GeneratorResumeModeThrow;
+    try {
+        for (;;) {
+            var result = this.evaluate(key, moduleRecord, fetcher, sentValue, resumeMode);
+            if (@getAbstractModuleRecordInternalField(moduleRecord, @abstractModuleRecordFieldState) === @GeneratorStateExecuting)
+                break;
+
+            try {
+                sentValue = await result;
+                resumeMode = @GeneratorResumeModeNormal;
+            } catch (error) {
+                sentValue = error;
+                resumeMode = @GeneratorResumeModeThrow;
+            }
+        }
+    } catch (error) {
+        @asyncModuleExecutionRejected(entry, error);
+        return;
+    }
+
+    this.asyncModuleExecutionFulfilled(entry, fetcher);
+}
+
+// https://tc39.es/ecma262/#sec-gather-available-ancestors
+@linkTimeConstant
+function gatherAvailableAncestors(entry, execList)
+{
+    for (var i = 0, { length } = entry.asyncParentModules; i < length; i++) {
+        var parentEntry = entry.asyncParentModules[i];
+        if (!execList.@includes(parentEntry) && parentEntry.cycleRoot.evaluationError === null) {
+            @assert(parentEntry.status === @ModuleStatusEvaluatingAsync);
+            @assert(parentEntry.evaluationError === null);
+            @assert(!!parentEntry.asyncEvaluation);
+            @assert(parentEntry.pendingAsyncDependencies > 0);
+
+            parentEntry.pendingAsyncDependencies--;
+            if (parentEntry.pendingAsyncDependencies === 0) {
+                @arrayPush(execList, parentEntry);
+                if (!parentEntry.hasTLA)
+                    @gatherAvailableAncestors(parentEntry, execList);
+            }
+        }
+    }
+}
+
+// https://tc39.es/ecma262/#sec-async-module-execution-fulfilled
+@visibility=PrivateRecursive
+function asyncModuleExecutionFulfilled(entry, fetcher)
+{
+    if (entry.status === @ModuleStatusEvaluated) {
+        @assert(entry.evaluationError !== null);
+        return;
+    }
+
+    @assert(entry.status === @ModuleStatusEvaluatingAsync);
+    @assert(!!entry.asyncEvaluation);
+    @assert(entry.evaluationError === null);
+
+    // `asyncEvaluation` isn't reset per https://github.com/tc39/ecma262/pull/2824
+    entry.status = @ModuleStatusEvaluated;
+
+    if (entry.topLevelCapability !== null) {
+        @assert(entry.cycleRoot === entry);
+        entry.topLevelCapability.resolve();
+    }
+
+    var execList = [];
+    @gatherAvailableAncestors(entry, execList);
+    execList.@sort((a, b) => a.asyncEvaluation - b.asyncEvaluation);
+
+    @assert(execList.@every(e => e.asyncEvaluation));
+    @assert(execList.@every(e => e.pendingAsyncDependencies === 0));
+    @assert(execList.@every(e => e.evaluationError === null));
```

Comment:
Do not use them.

Reply:
Replaced with inline function + old-school `for` loop.